src/services/services.ts

```diff
@@ -5295,6 +5308,165 @@ namespace ts {
             return getDefinitionFromSymbol(type.symbol, node);
         }
 
+         /// Goto implementation
+        function getImplementationAtPosition(fileName: string, position: number): ImplementationLocation[] {
+            synchronizeHostData();
+
+            const entries: ImplementationLocation[] = [];
+            const node = getTouchingPropertyName(getValidSourceFile(fileName), position);
+            const typeChecker = program.getTypeChecker();
+
+            // If invoked directly on a shorthand property assignment, then return
+            // the declaration of the symbol being assigned (not the symbol being assigned to).
+            if (node.parent.kind === SyntaxKind.ShorthandPropertyAssignment) {
+                const entry = getReferenceEntryForShorthandPropertyAssignment(node, typeChecker);
+                entries.push({
+                    textSpan: entry.textSpan,
+                    fileName: entry.fileName
+                });
+            }
+
+            // For most symbols, the definition is the same as the implementation so we can just
+            // call "Go to Definition". This case should handle anything that is not a type
+            // reference to or member of an interface, class, or union/intersection type.
+            else if (definitionIsImplementation(node, typeChecker)) {
+                const definitions = getDefinitionAtPosition(fileName, position);
+                forEach(definitions, (definition: DefinitionInfo) => {
+                    entries.push({
+                        textSpan: definition.textSpan,
+                        fileName: definition.fileName
+                    });
+                });
+            }
+
+            // Interfaces, classes, and unions/intersection types separate the implementation and
+            // definition so "Go to Definition" is not sufficient. This case handles invocations
+            // on type references and members of those types.
+            else {
+                // Perform "Find all References" and filter them down to implementations only
+                const result = getReferencedSymbolsForNode(node, program.getSourceFiles(), /*findInStrings*/false, /*findInComments*/false, /*implementations*/true);
+
+                forEach(result, referencedSymbol => {
+                    if (referencedSymbol.references) {
+                        forEach(referencedSymbol.references, entry => {
+                            entries.push({
+                                textSpan: entry.textSpan,
+                                fileName: entry.fileName
+                            });
+                        });
+                    }
+                });
+            }
+
+            return entries;
+        }
+
+        /**
+         * Returns true if the implementation for this node is the same as its definition
+         */
+        function definitionIsImplementation(node: Node, typeChecker: TypeChecker) {

```

Comment:
consider not special casing any of these and letting the default code path take care of them.